pallets/messages/src/lib.rs

```diff
@@ -315,28 +319,26 @@ impl<T: Config> Pallet<T> {
 		msa_id: Option<MessageSourceId>,
 		payload: BoundedVec<u8, T::MaxMessagePayloadSizeBytes>,
 		schema_id: SchemaId,
-	) -> Result<Message<T::MaxMessagePayloadSizeBytes>, DispatchError> {
-		<BlockMessages<T>>::try_mutate(
-			|existing_messages| -> Result<Message<T::MaxMessagePayloadSizeBytes>, DispatchError> {
-				let current_size: u16 = existing_messages
-					.len()
-					.try_into()
-					.map_err(|_| Error::<T>::TypeConversionOverflow)?;
-
-				let msg = Message {
-					payload, // size is checked on top of extrinsic
-					provider_msa_id,
-					msa_id,
-					index: current_size,
-				};
-
-				existing_messages
-					.try_push((msg.clone(), schema_id))
-					.map_err(|_| Error::<T>::TooManyMessagesInBlock)?;
-
-				Ok(msg)
-			},
-		)
+		current_block: T::BlockNumber,
+	) -> DispatchResult {
+		<Messages<T>>::try_mutate(current_block, schema_id, |existing_messages| -> DispatchResult {
+			let extrinsic_index =
+				<frame_system::Pallet<T>>::extrinsic_index().ok_or(Error::<T>::BadContext)?;
+			let msg = Message {
+				payload, // size is checked on top of extrinsic
+				provider_msa_id,
+				msa_id,
+				// converting u32 to u16 and defaulting to u16:MAX if overflowed
+				index: extrinsic_index.try_into().unwrap_or(u16::MAX),
+			};
+
+			existing_messages
+				.try_push(msg)
+				.map_err(|_| Error::<T>::TooManyMessagesInBlock)?;
+
+			Self::deposit_event(Event::MessagesStored { schema_id, block_number: current_block });
```

Comment:
Since we are removing `on_initialize` it does seem like we will need to add an event per message. Open to discuss about this and what we should include in here.

Reply:
I recall some cheaper single block storage options that are cheap as they are not in the tree. If so, perhaps a flag so we emit only on the first one for each schema?

We lose count however. Perhaps a single block storage and then an emit `on_finalize` as a simple set of event emissions would be fine for an `on_finalize`?

Reply:
Seems that there is some deciding to do here. However, let's try to move the event out of this function.